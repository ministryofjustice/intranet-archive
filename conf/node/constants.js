import { parseSchduleString } from "./controllers/schedule.js";

export const ordinalNumber = process.env.ORDINAL_NUMBER;
export const port = 2000;
export const jwt = process.env.JWT;

/**
 * Access
 */

export const sharedSecret = process.env.INTRANET_ARCHIVE_SHARED_SECRET;

/**
 * S3
 */

export const s3Region = "eu-west-2";
export const s3BucketName = process.env.S3_BUCKET_NAME;
export const s3Credentials = process.env.AWS_ACCESS_KEY_ID &&
  process.env.AWS_SECRET_ACCESS_KEY && {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  };

/**
 * CloudFront
 */

export const cloudFrontKeysObject =
  process.env.AWS_CLOUDFRONT_PUBLIC_KEYS_OBJECT;
export const cloudFrontPublicKey = process.env.AWS_CLOUDFRONT_PUBLIC_KEY;
export const cloudFrontPrivateKey = process.env.AWS_CLOUDFRONT_PRIVATE_KEY;

/**
 * Options
 */

export const corsOptions = {
  methods: ["POST"],
  origin: [
    "http://spider.intranet.docker/",
    "https://dev-intranet-archive.apps.live.cloud-platform.service.justice.gov.uk/",
  ],
};

/**
 * Defaults
 */

export const defaultAgency = "hq";
export const defaultUrl = "https://intranet.justice.gov.uk/";

/**
 * Validation
 */

export const allowedTargetHosts = [
  "intranet.docker",
  "intranet.justice.gov.uk",
  "dev.intranet.justice.gov.uk",
  "staging.intranet.justice.gov.uk",
  "demo.intranet.justice.gov.uk",
];

export const allowedTargetAgencies =
  process.env.ALLOWED_AGENCIES?.split(",") ?? [];

/**
 * Schedule
 */

export const snapshotSchedule = parseSchduleString(
  process.env.SNAPSHOT_SCHEDULE,
);

/**
 * Httrack
 */

export const sensitiveFiles = [
  "cookies.txt", // Contains the JWT and CloudFront cookies.
  "hts-in_progress.lock", // Contains the JWT.
  "hts-log.txt", // Has the httrack command line arguments - this includes the JWT.
  `hts-cache/doit.log`, // Has the httrack command line arguments - this includes the JWT.
  `hts-cache/new.zip`,
];

/**
 * Intranet application
 */

export const heartbeatEndpoint = "auth/heartbeat";

/**
 * Index pages
 */

export const indexCss = `
  main{display:block}a{background-color:transparent;color:#337ab7;text-decoration:none;text-shadow:#fff 1px 0 10px;}
  .list-group-item,body{background-color:#fff}a:active,a:hover{outline:0}h1{font-size:2em;margin:.67em 0}
  *,:after,:before{-webkit-box-sizing:border-box;-moz-box-sizing:border-box;box-sizing:border-box}
  html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;font-size:10px;-webkit-tap-highlight-color:transparent}
  body{margin:0;font-family:"Helvetica Neue",Helvetica,Arial,sans-serif;font-size:14px;line-height:1.42857143;color:#333}
  a:focus,a:hover{color:#23527c;text-decoration:underline}
  a:focus{outline:-webkit-focus-ring-color auto 5px;outline-offset:-2px}
  .container{padding-right:15px;padding-left:15px;margin-right:auto;margin-left:auto}
  @media (min-width:768px){.container{width:750px}}@media (min-width:992px){.container{width:970px}}
  @media (min-width:1200px){.container{width:1170px}}
  .list-group{padding-left:0;margin-bottom:20px;display:grid;grid-template-columns:repeat(auto-fill,minmax(160px,1fr))}
  .list-group-item{position:relative;display:block;padding:14px 18px;margin:0 20px 25px 0;background: #fff;
    -webkit-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    -moz-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
  }
  .container:after,.container:before{display:table;content:" "}
  .container:after{clear:both}.list-group{-webkit-box-shadow:0 1px 2px rgba(0,0,0,.075);box-shadow:0 1px 2px rgba(0,0,0,.075)}
`;
