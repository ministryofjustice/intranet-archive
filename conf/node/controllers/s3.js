import fs from "fs/promises";
import {
  S3Client,
  ListObjectsV2Command,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { S3SyncClient } from "s3-sync-client";
import mime from "mime-types";

import {
  s3BucketName,
  s3Credentials as credentials,
  s3Region,
  heartbeatEndpoint,
} from "../constants.js";

/**
 * S3 client options
 *
 * @type {import("@aws-sdk/client-s3").S3ClientConfig}
 */

export const s3Options = {
  region: s3Region,
  // Add the credentials if they exist - used for minio locally.
  ...(credentials && { credentials }),
  // Add the endpoint if it exists - used for minio locally.
  ...(process.env.S3_ENDPOINT && {
    endpoint: process.env.S3_ENDPOINT,
    forcePathStyle: true,
  }),
};

/**
 * S3 client
 *
 * @type {S3Client}
 */

export const client = new S3Client(s3Options);

/**
 * Create dummy /auth/heartbeat at bucket root, if it doesn't exist.
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @returns {Promise<void>}
 *
 * @throws {Error}
 */

export const createHeartbeat = async (
  bucket = s3BucketName,
  file = heartbeatEndpoint,
) => {
  const objects = await client.send(
    new ListObjectsV2Command({
      Bucket: bucket,
      Prefix: file,
    }),
  );

  if (!objects.Contents?.length) {
    const response = await client.send(
      new PutObjectCommand({
        Bucket: bucket,
        Key: file,
        Body: "OK",
      }),
    );
  }

  return;
};

/**
 * S3 sync client
 *
 * @see https://github.com/jeanbmar/s3-sync-client
 */

const { sync: originalSync } = new S3SyncClient({ client });

/**
 * Sync a local directory to an S3 bucket
 *
 * @param {string} source - The source directory
 * @param {string} destination - The destination directory
 * @param {?import('s3-sync-client/dist/commands/SyncCommand').SyncOptions} options - The sync options
 *
 * @returns {Promise<import('s3-sync-client/dist/commands/SyncCommand').SyncCommandOutput>}
 */

export const sync = async (source, destination, options = {}) => {
  // Set the content type for each file
  options.commandInput = (input) => ({
    ContentType: mime.lookup(input.Key) || "text/html",
  });

  return originalSync(source, destination, options);
};

/**
 * Empty an S3 folder by using sync and deleting all files
 *
 * @param {string} path - The path to empty
 */

export const s3EmptyDir = async (path) => {
  // Make a tmp empty directory
  const emptyDir = `/tmp/${Date.now()}`;

  // Ensure the directory exists
  await fs.mkdir(emptyDir, { recursive: true });

  // Sync the empty directory to the folder
  await sync(emptyDir, `s3://${s3BucketName}/${path}`, { del: true });

  // Remove the empty directory
  await fs.rm(emptyDir, { recursive: true });
};

/**
 * Check if the bucket is accessible by listing the root of the bucket
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @returns {Promise<boolean>}
 *
 * @throws {Error}
 */

export const checkAccess = async (bucket = s3BucketName) => {
  const command = new ListObjectsV2Command({
    Bucket: bucket,
  });

  await client.send(command);

  return true;
};

/**
 * Returns an array of all the agencies in the bucket's host folder.
 *
 * This function looks in the bucket in the host folder and returns an array of all the agencies.
 * e.g looks in s3://s3BucketName/intranet.justice.gov.uk/ and returns an array of hq,hmcts etc.
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @param {string} host - The host to the intranet e.g. intranet.justice.gov.uk or dev.intranet.justice.gov.uk
 * @returns {Promise<string[]>}
 *
 * @throws {Error}
 */

export const getAgenciesFromS3 = async (bucket = s3BucketName, host) => {
  if (!host) {
    throw new Error("Host is required");
  }

  const command = new ListObjectsV2Command({
    Bucket: bucket,
    Prefix: `${host}/`,
    Delimiter: "/",
  });

  const { CommonPrefixes } = await client.send(command);

  return CommonPrefixes.map((folder) =>
    folder.Prefix.replace(`${host}/`, "").replace("/", ""),
  );
};

/**
 * Get an agencies snapshots from S3
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @param {string} host - The host to the intranet e.g. intranet.justice.gov.uk or dev.intranet.justice.gov.uk
 * @param {string} agency - The agency to get snapshots for e.g. hq, hmcts etc.
 * @returns {Promise<string[]>}
 *
 * @throws {Error}
 */

export const getSnapshotsFromS3 = async (
  bucket = s3BucketName,
  host,
  agency,
) => {
  if (!host) {
    throw new Error("Host is required");
  }

  if (!agency) {
    throw new Error("Agency is required");
  }

  const command = new ListObjectsV2Command({
    Bucket: bucket,
    Prefix: `${host}/${agency}/`,
    Delimiter: "/",
  });

  const { CommonPrefixes } = await client.send(command);

  return CommonPrefixes.map(({ Prefix }) =>
    // Remove the host and agency from the Prefix
    Prefix.replace(`${host}/${agency}/`, "").replace("/", ""),
  ).filter((folder) =>
    // Filter out any folders that are not in the format YYYY-MM-DD
    /^\d{4}-\d{2}-\d{2}$/.test(folder),
  );
};

/**
 * Write string to an S3 file.
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @param {string} path - The path to write to
 * @param {string} content - The content to write
 * @param {Object} [options] - The options object
 * @param {number} [options.cacheMaxAge] - The cache max age in seconds - a simpler alternative to invalidating CloudFront cache
 *
 * @returns {Promise<void>}
 */

export const writeToS3 = async (
  bucket = s3BucketName,
  path,
  content,
  { cacheMaxAge } = {},
) => {
  const command = new PutObjectCommand({
    Bucket: bucket,
    Key: path,
    Body: content,
    ContentType: mime.lookup(path) || "text/html",
    ...(cacheMaxAge && { CacheControl: `max-age=${cacheMaxAge}` }),
  });

  await client.send(command);
};
