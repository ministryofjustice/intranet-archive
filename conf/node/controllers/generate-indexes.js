import { intranetUrls, s3BucketName, indexCss } from "../constants.js";
import { getEnvironmentIndex, getAgencyPath } from "./paths.js";
import { getAgenciesFromS3, getSnapshotsFromS3 } from "./s3.js";

/**
 * Generate the root index html
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @param {string} env - The env of the intranet e.g. production or dev.
 * @returns {Promise<string>}
 *
 * @throws {Error}
 */

export const generateRootIndex = async (bucket = s3BucketName, env) => {
  if (!env) {
    throw new Error("Env is required");
  }

  const agencies = await getAgenciesFromS3(bucket, env);

  const hostname = new URL(intranetUrls[env]).hostname;

  // Get any other envs that have agencies, so that we can link to them.
  // This will only be shown on dev, so the presentation is not a priority.
  const otherLinks = (
    await Promise.all(
      Object.keys(intranetUrls)
        // Remove the current env
        .filter((e) => e !== env)
        // Check if each env has agencies and map to an object with the href and label
        .map(async (e) => ({
          hasAgencies: (await getAgenciesFromS3(bucket, e))?.length > 0,
          href: `/${getEnvironmentIndex(e)}`,
          label: `${e}`,
        })),
    )
  )
    // Only show links to envs with agencies
    .filter(({ hasAgencies }) => hasAgencies);

  const html = `<!doctype html><html lang="en">
    <head><title>Intranet Archive Index</title><style>${indexCss}</style></head>
    <body>
      <main>
        <div class="container container--flex">
          <h1>Ministry of Justice Intranet Archive</h1>
          ${
            otherLinks.length > 0
              ? `<nav class="top-links">
                  Switch target environment: 
                  ${otherLinks
                    .map((link) => `<a href="${link.href}" >${link.label}</a>`)
                    .join(" | ")}
                </nav>`
              : ``
          }
        </div>
        <div class="container">
          <h2>${hostname}</h2>
          <ul class="list-group">
            ${agencies
              .map(
                (agency) => `
                <li class="list-group-item">
                  <a href="/${getAgencyPath(
                    env,
                    agency,
                  )}/index.html">${agency}</a>
                </li>`,
              )
              .join("\n")}
          </ul>
        </div>
      </main>
    </body>
  </html>`;

  return html;
};

/**
 * Generate the agency index html
 *
 * @param {string} bucket - The bucket name, defaults to the s3BucketName constant
 * @param {string} env - The environment for the intranet e.g. production or dev
 * @param {string} agency - The agency to get snapshots for e.g. hq, hmcts etc.
 * @returns {Promise<string>}
 *
 * @throws {Error}
 */

export const generateAgencyIndex = async (
  bucket = s3BucketName,
  env,
  agency,
) => {
  if (!env) {
    throw new Error("Env is required");
  }

  if (!agency) {
    throw new Error("Agency is required");
  }

  const url = new URL(intranetUrls[env]);

  const snapshots = await getSnapshotsFromS3(bucket, env, agency);

  const html = `<!doctype html><html lang="en">
    <head><title>Intranet Archive Index</title><style>${indexCss}</style></head>
    <body>
      <main>
        <div class="container container--flex">
          <h1>Ministry of Justice Intranet Archive</h1>
          <nav class="top-links">
            <a href="/${getEnvironmentIndex(env)}">Switch to other intranet</a>
          </nav>
        </div>
        <div class="container">
          <h2>${url.hostname} - ${agency}</h2>
          <ul class="list-group">
            ${snapshots
              .map(
                (snapshot) => `
                <li class="list-group-item">
                  <a href="/${getAgencyPath(env, agency)}/${snapshot}/${
                  url.hostname
                }/index.html" target="_blank">${snapshot}</a>
                </li>`,
              )
              .join("\n")}
          </ul>
        </div>
      </main>
    </body>
  </html>`;

  return html;
};
