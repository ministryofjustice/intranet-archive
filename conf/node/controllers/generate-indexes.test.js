import {
  S3Client,
  DeleteObjectCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";

import { s3BucketName } from "../constants.js";
import { generateRootIndex, generateAgencyIndex } from "./generate-indexes.js";

import { s3Options } from "./s3.js";

let client;

beforeAll(async () => {
  client = new S3Client(s3Options);

  // Make a folder on s3 for the test
  await client.send(
    new PutObjectCommand({
      Bucket: s3BucketName,
      Key: "dev-hmcts/2024-01-01/index.html",
      Body: "test",
    }),
  );

  await client.send(
    new PutObjectCommand({
      Bucket: s3BucketName,
      Key: "dev-hq/2024-01-01/index.html",
      Body: "test",
    }),
  );
});

afterAll(() => {
  // Remove the test folder
  client.send(
    new DeleteObjectCommand({
      Bucket: s3BucketName,
      Key: "dev-hmcts/",
    }),
  );

  client.send(
    new DeleteObjectCommand({
      Bucket: s3BucketName,
      Key: "dev-hq/",
    }),
  );

  client.destroy();
});

describe("generateRootIndex", () => {
  it("should throw an error if the env is not provided", async () => {
    await expect(generateRootIndex(undefined, undefined)).rejects.toThrow(
      "Env is required",
    );
  });

  it("should return the agencies as a string", async () => {
    const result = await generateRootIndex(
      s3BucketName,
      "dev",
    );
    expect(result).toMatchSnapshot();
  });
});

describe("generateAgencyIndex", () => {
  it("should throw an error if the env is not provided", async () => {
    await expect(generateAgencyIndex(undefined, undefined, undefined)).rejects.toThrow(
      "Env is required",
    );
  });

  it("should throw an error if the agency is not provided", async () => {
    await expect(generateAgencyIndex(undefined, "dev", undefined)).rejects.toThrow(
      "Agency is required",
    );
  });

  it("should return the snapshots as a string", async () => {
    const result = await generateAgencyIndex(
      s3BucketName,
      "dev",
      "hmcts",
    );
    expect(result).toMatchSnapshot();
  });
});
