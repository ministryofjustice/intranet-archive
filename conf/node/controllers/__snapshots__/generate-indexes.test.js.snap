// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAgencyIndex should return the snapshots as a string 1`] = `
"<!doctype html><html lang="en">
    <head><title>Intranet Archive Index</title><style>
  main{display:block}a{background-color:transparent;color:#337ab7;text-decoration:none;text-shadow:#fff 1px 0 10px;}
  .list-group-item,body{background-color:#fff}a:active,a:hover{outline:0}h1{font-size:2em;margin:.67em 0}
  *,:after,:before{-webkit-box-sizing:border-box;-moz-box-sizing:border-box;box-sizing:border-box}
  html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;font-size:10px;-webkit-tap-highlight-color:transparent}
  body{margin:0;font-family:"Helvetica Neue",Helvetica,Arial,sans-serif;font-size:14px;line-height:1.42857143;color:#333}
  a:focus,a:hover{color:#23527c;text-decoration:underline}
  a:focus{outline:-webkit-focus-ring-color auto 5px;outline-offset:-2px}
  .container{padding-right:15px;padding-left:15px;margin-right:auto;margin-left:auto}
  @media (min-width:768px){.container{width:750px}}@media (min-width:992px){.container{width:970px}}
  @media (min-width:1200px){.container{width:1170px}}
  .list-group{padding-left:0;margin-bottom:20px;display:grid;grid-template-columns:repeat(auto-fill,minmax(160px,1fr))}
  .list-group-item{position:relative;display:block;padding:14px 18px;margin:0 20px 25px 0;background: #fff;
    -webkit-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    -moz-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
  }
  .container:after,.container:before{display:table;content:" "}
  .container:after{clear:both}.list-group{-webkit-box-shadow:0 1px 2px rgba(0,0,0,.075);box-shadow:0 1px 2px rgba(0,0,0,.075)}
</style></head>
    <body>
      <main>
        <div class="container px-4 py-5">
          <h1>Ministry of Justice Intranet Archive</h1>
          <h2 class="pb-2 border-bottom">test.generate.indexes - hmcts</h2>
          <ul class="list-group">
            
                <li class="list-group-item">
                  <a href="/test.generate.indexes/hmcts/2024-01-01/test.generate.indexes/index.html" target="_blank">2024-01-01</a>
                </li>
          </ul>
        </div>
      </main>
    </body>
  </html>"
`;

exports[`generateRootIndex should return the agencies as a string 1`] = `
"<!doctype html><html lang="en">
    <head><title>Intranet Archive Index</title><style>
  main{display:block}a{background-color:transparent;color:#337ab7;text-decoration:none;text-shadow:#fff 1px 0 10px;}
  .list-group-item,body{background-color:#fff}a:active,a:hover{outline:0}h1{font-size:2em;margin:.67em 0}
  *,:after,:before{-webkit-box-sizing:border-box;-moz-box-sizing:border-box;box-sizing:border-box}
  html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;font-size:10px;-webkit-tap-highlight-color:transparent}
  body{margin:0;font-family:"Helvetica Neue",Helvetica,Arial,sans-serif;font-size:14px;line-height:1.42857143;color:#333}
  a:focus,a:hover{color:#23527c;text-decoration:underline}
  a:focus{outline:-webkit-focus-ring-color auto 5px;outline-offset:-2px}
  .container{padding-right:15px;padding-left:15px;margin-right:auto;margin-left:auto}
  @media (min-width:768px){.container{width:750px}}@media (min-width:992px){.container{width:970px}}
  @media (min-width:1200px){.container{width:1170px}}
  .list-group{padding-left:0;margin-bottom:20px;display:grid;grid-template-columns:repeat(auto-fill,minmax(160px,1fr))}
  .list-group-item{position:relative;display:block;padding:14px 18px;margin:0 20px 25px 0;background: #fff;
    -webkit-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    -moz-box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
    box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
  }
  .container:after,.container:before{display:table;content:" "}
  .container:after{clear:both}.list-group{-webkit-box-shadow:0 1px 2px rgba(0,0,0,.075);box-shadow:0 1px 2px rgba(0,0,0,.075)}
</style></head>
    <body>
      <main>
        <div class="container px-4 py-5">
          <h1>Ministry of Justice Intranet Archive</h1>
          <h2 class="pb-2 border-bottom">test.generate.indexes</h2>
          <ul class="list-group">
            
                <li class="list-group-item">
                  <a href="/test.generate.indexes/hmcts/index.html" target="_blank">hmcts</a>
                </li>

                <li class="list-group-item">
                  <a href="/test.generate.indexes/hq/index.html" target="_blank">hq</a>
                </li>
          </ul>
        </div>
      </main>
    </body>
  </html>"
`;
